src/radical/pilot/agent/agent_0.py

```diff
@@ -340,77 +353,130 @@ def _start_services(self):
+        #     'rank_per_node'     : 2,
+            # 'task_sandbox_path' : self._pwd,
+            # 'description'       : TaskDescription({
+            #     'uid'           : service_task_uid,
+            #     'ranks'         : 1,
+            #     'cores_per_rank': self._rm.info.cores_per_node,
+            #     'executable'    : '/bin/sh',
+            #     'arguments'     : [bs_name, 'services']
+            # }).as_dict(),
+            # 'slots': {'ranks'   : [{'node_name': nodes[0]['node_name'],
+            #                         'node_id'  : nodes[0]['node_id'],
+            #                         'core_map' : [[0]],
+            #                         'gpu_map'  : [],
+            #                         'lfs'      : 0,
+            #                         'mem'      : 0}]}
+        # }
+
+        # launcher = self._rm.find_launcher(service_task)
+        # if not launcher:
+        #     raise RuntimeError('no launch method found for sub agent')
 
-        tmp  = '#!/bin/sh\n\n'
-        tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
-        cmds = launcher.get_launcher_env()
-        for cmd in cmds:
-            tmp += '%s || exit 1\n' % cmd
-        cmds = launcher.get_launch_cmd(service_task, ex_name)
-        tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
+        # FIXME: set RP environment (as in Popen Executor)
 
-        with ru.ru_open(ls_name, 'w') as fout:
-            fout.write(tmp)
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
+        # cmds = launcher.get_launcher_env()
+        # for cmd in cmds:
+        #     tmp += '%s || exit 1\n' % cmd
+        # cmds = launcher.get_launch_cmd(service_task, ex_name)
+        # tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
+        #
+        # with ru.ru_open(ls_name, 'w') as fout:
+        #     fout.write(tmp)
 
-        tmp  = '#!/bin/sh\n\n'
-        tmp += '. ./env/service.env\n'
-        tmp += '/bin/sh -l ./services\n\n'
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += '. ./env/service.env\n'
+        # tmp += '/bin/sh -l ./services\n\n'
+        #
+        # with ru.ru_open(ex_name, 'w') as fout:
+        #     fout.write(tmp)
+        #
+        #
+        # make sure scripts are executable
+        # st_l = os.stat(ls_name)
+        # st_e = os.stat(ex_name)
+        # os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
+        # os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
+        #
+        # spawn the sub-agent
+        # cmdline = './%s' % ls_name
+        #
+        # self._log.info('create services: %s' % cmdline)
+        # ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
 
-        with ru.ru_open(ex_name, 'w') as fout:
-            fout.write(tmp)
+        # self.publish(rpc.STATE_PUBSUB, {'cmd': 'service_state_update','arg': services})
 
+    def _state_cb_of_services(self, topic, msg):
+        cmd = msg['cmd']
+        services = msg['arg']
```

Comment:
You do not yet know if the state update is for your service - it could be for any other task.  You check if it is service task only on line 470 below.  So maybe `tasks = msg['arg']` is more appropriate?

It's a minor thing, but makes the code more readable.

Reply:
Yes definitely.